File: tools/airc/src/interp.ts
import type { Expr, FnItem, Module } from "./ast.js";

export type AirValue =
  | { tag: "i32"; v: number }
  | { tag: "bool"; v: boolean }
  | { tag: "str"; v: string }
  | { tag: "ok"; v: AirValue }
  | { tag: "err"; v: AirValue }
  | { tag: "array"; elems: AirValue[] };

/** When set, `cap.print` appends here instead of writing to console. */
let stdoutCapture: string[] | null = null;

/** Run `fn` while capturing lines from `cap.print` (not CLI return-value echo). */
export function withStdoutCapture<T>(fn: () => T): { result: T; lines: string[] } {
  stdoutCapture = [];
  try {
    const result = fn();
    return { result, lines: stdoutCapture ?? [] };
  } finally {
    stdoutCapture = null;
  }
}

function hostPrintLine(line: string): void {
  if (stdoutCapture) stdoutCapture.push(line);
  else console.log(line);
}

class BreakSignal {
  constructor(public value: AirValue) {}
}

function litValue(width: string, digits: string): AirValue {
  if (width === "bool") return { tag: "bool", v: digits === "true" };
  if (width === "str") return { tag: "str", v: digits };
  if (width.startsWith("f")) return { tag: "i32", v: Number(digits) }; // Phase1: store floats as number in i32 tag loosely — use i32 path for ints only in examples
  const n = Number(digits);
  return { tag: "i32", v: n | 0 };
}

function asI32(v: AirValue): number {
  if (v.tag !== "i32") throw new Error(`runtime.type: expected i32 got ${v.tag}`);
  return v.v;
}

function evalExpr(e: Expr, env: Map<string, AirValue>, fns: Map<string, FnItem>): AirValue {
  if (typeof e === "boolean") return { tag: "bool", v: e };
  if (typeof e === "number") return { tag: "i32", v: e | 0 };
  if (typeof e === "string") return { tag: "str", v: e };

  switch (e[0]) {
    case "lit":
      return litValue(e[1], e[2]);
    case "var": {
      const v = env.get(e[1]);
      if (!v) throw new Error(`runtime.unbound: ${e[1]}`);
      return v;
    }
    case "seq": {
      let last: AirValue = { tag: "i32", v: 0 };
      for (let i = 1; i < e.length; i++) last = evalExpr(e[i] as Expr, env, fns);
      return last;
    }
    case "let": {
      const child = new Map(env);
      for (const [name, , init] of e[1]) {
        child.set(name, evalExpr(init, child, fns));
      }
      return evalExpr(e[2], child, fns);
    }
    case "set!": {
      const v = evalExpr(e[2], env, fns);
      env.set(e[1], v);
      return v;
    }
    case "if": {
      const c = evalExpr(e[1], env, fns);
      if (c.tag !== "bool") throw new Error("runtime.type: if cond");
      return evalExpr(c.v ? e[2] : e[3], env, fns);
    }
    case "loop": {
      for (;;) {
        try {
          evalExpr(e[1], env, fns);
        } catch (ex) {
          if (ex instanceof BreakSignal) return ex.value;
          throw ex;
        }
      }
    }
    case "break":
      throw new BreakSignal(evalExpr(e[1], env, fns));
    case "return":
      return evalExpr(e[1], env, fns);
    case "call": {
      const callee = e[1];
      if (typeof callee !== "string") throw new Error("runtime: callee");
      if (callee === "aset") {
        const place = e[2] as Expr;
        if (!Array.isArray(place) || place[0] !== "var" || typeof place[1] !== "string") {
          throw new Error('runtime.aset: place must be ["var", name]');
        }
        const name = place[1];
        const idx = asI32(evalExpr(e[3] as Expr, env, fns));
        const val = evalExpr(e[4] as Expr, env, fns);
        const slot = env.get(name);
        if (!slot || slot.tag !== "array") throw new Error("runtime.aset");
        if (idx < 0 || idx >= slot.elems.length) throw new Error("runtime.oob");
        slot.elems[idx] = val;
        return { tag: "i32", v: 0 };
      }
      const args = (e.slice(2) as Expr[]).map((a) => evalExpr(a, env, fns));
      if (callee === "+") return { tag: "i32", v: (asI32(args[0]!) + asI32(args[1]!)) | 0 };
      if (callee === "-") return { tag: "i32", v: (asI32(args[0]!) - asI32(args[1]!)) | 0 };
      if (callee === "*") return { tag: "i32", v: Math.imul(asI32(args[0]!), asI32(args[1]!)) };
      if (callee === "/") {
        const b = asI32(args[1]!);
        if (b === 0) throw new Error("runtime.div0");
        return { tag: "i32", v: (asI32(args[0]!) / b) | 0 };
      }
      if (callee === "%") {
        const b = asI32(args[1]!);
        if (b === 0) throw new Error("runtime.div0");
        return { tag: "i32", v: asI32(args[0]!) % b | 0 };
      }
      if (callee === "<") return { tag: "bool", v: asI32(args[0]!) < asI32(args[1]!) };
      if (callee === "<=") return { tag: "bool", v: asI32(args[0]!) <= asI32(args[1]!) };
      if (callee === ">") return { tag: "bool", v: asI32(args[0]!) > asI32(args[1]!) };
      if (callee === ">=") return { tag: "bool", v: asI32(args[0]!) >= asI32(args[1]!) };
      if (callee === "==") {
        const a = args[0]!;
        const b = args[1]!;
        if (a.tag === "i32" && b.tag === "i32") return { tag: "bool", v: a.v === b.v };
        if (a.tag === "bool" && b.tag === "bool") return { tag: "bool", v: a.v === b.v };
        if (a.tag === "str" && b.tag === "str") return { tag: "bool", v: a.v === b.v };
        return { tag: "bool", v: false };
      }
      if (callee === "!=") {
        const eq = evalExpr(["call", "==", e[2] as Expr, e[3] as Expr], env, fns);
        return { tag: "bool", v: !(eq as { tag: "bool"; v: boolean }).v };
      }
      if (callee === "ok") return { tag: "ok", v: args[0]! };
      if (callee === "err") return { tag: "err", v: args[0]! };
      if (
        callee === "checked_add" ||
        callee === "checked_sub" ||
        callee === "checked_mul" ||
        callee === "checked_div"
      ) {
        const a = asI32(args[0]!);
        const b = asI32(args[1]!);
        if (callee === "checked_div") {
          if (b === 0) return { tag: "err", v: { tag: "str", v: "div0" } };
          if (a === -2147483648 && b === -1) {
            return { tag: "err", v: { tag: "str", v: "overflow" } };
          }
          return { tag: "ok", v: { tag: "i32", v: (a / b) | 0 } };
        }
        let wide: bigint;
        if (callee === "checked_add") wide = BigInt(a) + BigInt(b);
        else if (callee === "checked_sub") wide = BigInt(a) - BigInt(b);
        else wide = BigInt(a) * BigInt(b);
        if (wide > 2147483647n || wide < -2147483648n) {
          return { tag: "err", v: { tag: "str", v: "overflow" } };
        }
        return { tag: "ok", v: { tag: "i32", v: Number(wide) } };
      }
      if (callee === "aget") {
        const arr = args[0]!;
        const idx = asI32(args[1]!);
        if (arr.tag !== "array") throw new Error("runtime.aget");
        if (idx < 0 || idx >= arr.elems.length) throw new Error("runtime.oob");
        return arr.elems[idx]!;
      }
      const fn = fns.get(callee);
      if (!fn) throw new Error(`runtime.unbound fn ${callee}`);
      const frame = new Map<string, AirValue>();
      fn[2].forEach(([name], i) => frame.set(name, args[i]!));
      return evalExpr(fn[4], frame, fns);
    }
    case "as":
      return evalExpr(e[2], env, fns);
    case "match": {
      const scr = evalExpr(e[1], env, fns);
      for (let i = 2; i < e.length; i++) {
        const [pat, body] = e[i] as [unknown, Expr];
        const child = new Map(env);
        if (Array.isArray(pat) && pat[0] === "ok" && scr.tag === "ok" && typeof pat[1] === "string") {
          child.set(pat[1], scr.v);
          return evalExpr(body, child, fns);
        }
        if (Array.isArray(pat) && pat[0] === "err" && scr.tag === "err" && typeof pat[1] === "string") {
          child.set(pat[1], scr.v);
          return evalExpr(body, child, fns);
        }
      }
      throw new Error("runtime.match");
    }
    case "array_lit": {
      const elems = (e.slice(2) as Expr[]).map((x) => evalExpr(x, env, fns));
      return { tag: "array", elems };
    }
    case "cap": {
      if (e[1] === "print") {
        const v = evalExpr(e[2] as Expr, env, fns);
        if (v.tag === "str") hostPrintLine(v.v);
        else hostPrintLine(JSON.stringify(v));
      }
      return { tag: "i32", v: 0 };
    }
    default:
      throw new Error(`runtime.unsupported ${e[0]}`);
  }
}

export function runModule(mod: Module): AirValue {
  const fns = new Map<string, FnItem>();
  for (let i = 2; i < mod.length; i++) {
    const it = mod[i];
    if (Array.isArray(it) && it[0] === "fn") fns.set(it[1] as string, it as FnItem);
  }
  const main = fns.get("main");
  if (!main) throw new Error("runtime.main");
  return evalExpr(main[4], new Map(), fns);
}

export function valueToExitCode(v: AirValue): number {
  if (v.tag === "i32") return v.v & 0xff;
  return 0;
}

File: crates/airc/src/interp.rs
use crate::parse::{find_fn, Module};
use serde_json::Value;
use std::cell::RefCell;
use std::collections::HashMap;

thread_local! {
    static STDOUT_CAPTURE: RefCell<Option<Vec<String>>> = const { RefCell::new(None) };
}

/// Run `f` while capturing lines written by `cap.print` (not CLI `print_value`).
pub fn with_stdout_capture<R>(f: impl FnOnce() -> R) -> (R, Vec<String>) {
    STDOUT_CAPTURE.with(|c| {
        *c.borrow_mut() = Some(Vec::new());
    });
    let result = f();
    let lines = STDOUT_CAPTURE.with(|c| c.borrow_mut().take().unwrap_or_default());
    (result, lines)
}

fn host_print_line(line: &str) {
    STDOUT_CAPTURE.with(|c| {
        if let Some(buf) = c.borrow_mut().as_mut() {
            buf.push(line.to_string());
        } else {
            println!("{line}");
        }
    });
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AirValue {
    I32(i32),
    Bool(bool),
    Str(String),
    Ok(Box<AirValue>),
    Err(Box<AirValue>),
    Array(Vec<AirValue>),
}

#[derive(Debug)]
pub enum RuntimeError {
    Message(String),
}

impl std::fmt::Display for RuntimeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RuntimeError::Message(m) => write!(f, "{m}"),
        }
    }
}

#[derive(Debug)]
enum EvalErr {
    Break(AirValue),
    Msg(String),
}

fn tag<'a>(v: &'a Value) -> Option<(&'a str, &'a [Value])> {
    let arr = v.as_array()?;
    let t = arr.first()?.as_str()?;
    Some((t, &arr[1..]))
}

fn as_i32(v: &AirValue) -> Result<i32, EvalErr> {
    match v {
        AirValue::I32(n) => Ok(*n),
        _ => Err(EvalErr::Msg("runtime.type: expected i32".into())),
    }
}

fn lit_value(width: &str, digits: &str) -> Result<AirValue, EvalErr> {
    if width == "bool" {
        return Ok(AirValue::Bool(digits == "true"));
    }
    if width == "str" {
        return Ok(AirValue::Str(digits.to_string()));
    }
    let n: i32 = digits
        .parse()
        .map_err(|_| EvalErr::Msg(format!("runtime.lit: {digits}")))?;
    Ok(AirValue::I32(n))
}

fn eval2(
    e: &Value,
    env: &mut HashMap<String, AirValue>,
    fns: &HashMap<String, &Value>,
) -> Result<AirValue, EvalErr> {
    if let Some(b) = e.as_bool() {
        return Ok(AirValue::Bool(b));
    }
    if let Some(n) = e.as_i64() {
        return Ok(AirValue::I32(n as i32));
    }
    if let Some(s) = e.as_str() {
        return Ok(AirValue::Str(s.to_string()));
    }
    let (t, rest) = tag(e).ok_or_else(|| EvalErr::Msg("runtime.expr".into()))?;
    match t {
        "lit" => lit_value(rest[0].as_str().unwrap(), rest[1].as_str().unwrap()),
        "var" => {
            let name = rest[0].as_str().unwrap();
            env.get(name)
                .cloned()
                .ok_or_else(|| EvalErr::Msg(format!("runtime.unbound: {name}")))
        }
        "seq" => {
            let mut last = AirValue::I32(0);
            for x in rest {
                last = eval2(x, env, fns)?;
            }
            Ok(last)
        }
        "let" => {
            let mut child = env.clone();
            for b in rest[0].as_array().unwrap() {
                let ba = b.as_array().unwrap();
                let name = ba[0].as_str().unwrap().to_string();
                let init = if ba.len() == 2 { &ba[1] } else { &ba[2] };
                let v = eval2(init, &mut child, fns)?;
                child.insert(name, v);
            }
            eval2(&rest[1], &mut child, fns)
        }
        "set!" => {
            let name = rest[0].as_str().unwrap().to_string();
            let v = eval2(&rest[1], env, fns)?;
            env.insert(name, v.clone());
            Ok(v)
        }
        "if" => match eval2(&rest[0], env, fns)? {
            AirValue::Bool(true) => eval2(&rest[1], env, fns),
            AirValue::Bool(false) => eval2(&rest[2], env, fns),
            _ => Err(EvalErr::Msg("runtime.type: if cond".into())),
        },
        "loop" => loop {
            match eval2(&rest[0], env, fns) {
                Err(EvalErr::Break(v)) => return Ok(v),
                Err(e) => return Err(e),
                Ok(_) => continue,
            }
        },
        "break" => Err(EvalErr::Break(eval2(&rest[0], env, fns)?)),
        "return" => eval2(&rest[0], env, fns),
        "call" => {
            let callee = rest[0].as_str().unwrap();
            let mut args = Vec::new();
            for a in &rest[1..] {
                args.push(eval2(a, env, fns)?);
            }
            match callee {
                "+" => Ok(AirValue::I32(
                    as_i32(&args[0])?.wrapping_add(as_i32(&args[1])?),
                )),
                "-" => Ok(AirValue::I32(
                    as_i32(&args[0])?.wrapping_sub(as_i32(&args[1])?),
                )),
                "*" => Ok(AirValue::I32(
                    as_i32(&args[0])?.wrapping_mul(as_i32(&args[1])?),
                )),
                "/" => {
                    let b = as_i32(&args[1])?;
                    if b == 0 {
                        return Err(EvalErr::Msg("runtime.div0".into()));
                    }
                    Ok(AirValue::I32(as_i32(&args[0])? / b))
                }
                "<=" => Ok(AirValue::Bool(as_i32(&args[0])? <= as_i32(&args[1])?)),
                "<" => Ok(AirValue::Bool(as_i32(&args[0])? < as_i32(&args[1])?)),
                ">" => Ok(AirValue::Bool(as_i32(&args[0])? > as_i32(&args[1])?)),
                ">=" => Ok(AirValue::Bool(as_i32(&args[0])? >= as_i32(&args[1])?)),
                "==" => match (&args[0], &args[1]) {
                    (AirValue::I32(a), AirValue::I32(b)) => Ok(AirValue::Bool(a == b)),
                    (AirValue::Bool(a), AirValue::Bool(b)) => Ok(AirValue::Bool(a == b)),
                    (AirValue::Str(a), AirValue::Str(b)) => Ok(AirValue::Bool(a == b)),
                    _ => Ok(AirValue::Bool(false)),
                },
                "ok" => Ok(AirValue::Ok(Box::new(args[0].clone()))),
                "err" => Ok(AirValue::Err(Box::new(args[0].clone()))),
                "aget" => {
                    let idx = as_i32(&args[1])?;
                    match &args[0] {
                        AirValue::Array(elems) => {
                            if idx < 0 || idx as usize >= elems.len() {
                                return Err(EvalErr::Msg("runtime.oob".into()));
                            }
                            Ok(elems[idx as usize].clone())
                        }
                        _ => Err(EvalErr::Msg("runtime.aget".into())),
                    }
                }
                other => {
                    let f = fns
                        .get(other)
                        .ok_or_else(|| EvalErr::Msg(format!("runtime.unbound fn {other}")))?;
                    let arr = f.as_array().unwrap();
                    let mut frame = HashMap::new();
                    for (i, p) in arr[2].as_array().unwrap().iter().enumerate() {
                        let name = p.as_array().unwrap()[0].as_str().unwrap().to_string();
                        frame.insert(name, args[i].clone());
                    }
                    eval2(&arr[4], &mut frame, fns)
                }
            }
        }
        "match" => {
            let scr = eval2(&rest[0], env, fns)?;
            for arm in &rest[1..] {
                let aa = arm.as_array().unwrap();
                let pat = aa[0].as_array().unwrap();
                let body = &aa[1];
                let mut child = env.clone();
                match (pat[0].as_str(), &scr) {
                    (Some("ok"), AirValue::Ok(v)) => {
                        let name = pat[1].as_str().unwrap().to_string();
                        child.insert(name, (**v).clone());
                        return eval2(body, &mut child, fns);
                    }
                    (Some("err"), AirValue::Err(v)) => {
                        let name = pat[1].as_str().unwrap().to_string();
                        child.insert(name, (**v).clone());
                        return eval2(body, &mut child, fns);
                    }
                    _ => continue,
                }
            }
            Err(EvalErr::Msg("runtime.match".into()))
        }
        "array_lit" => {
            let mut elems = Vec::new();
            for el in &rest[1..] {
                elems.push(eval2(el, env, fns)?);
            }
            Ok(AirValue::Array(elems))
        }
        "as" => eval2(&rest[1], env, fns),
        "cap" => {
            if rest[0].as_str() == Some("print") {
                let v = eval2(&rest[1], env, fns)?;
                match v {
                    AirValue::Str(s) => host_print_line(&s),
                    other => host_print_line(&format!("{other:?}")),
                }
            }
            Ok(AirValue::I32(0))
        }
        other => Err(EvalErr::Msg(format!("runtime.unsupported {other}"))),
    }
}

pub fn run_module(module: &Module) -> Result<AirValue, RuntimeError> {
    let mut fns = HashMap::new();
    for f in crate::parse::fns_in_module(module) {
        let name = f.as_array().unwrap()[1].as_str().unwrap().to_string();
        fns.insert(name, f);
    }
    let main =
        find_fn(module, "main").ok_or_else(|| RuntimeError::Message("runtime.main".into()))?;
    let arr = main.as_array().unwrap();
    let mut env = HashMap::new();
    match eval2(&arr[4], &mut env, &fns) {
        Ok(v) => Ok(v),
        Err(EvalErr::Break(v)) => Ok(v),
        Err(EvalErr::Msg(m)) => Err(RuntimeError::Message(m)),
    }
}

pub fn value_to_exit_code(v: &AirValue) -> u8 {
    match v {
        AirValue::I32(n) => *n as u8,
        _ => 0,
    }
}

pub fn print_value(v: &AirValue) {
    match v {
        AirValue::I32(n) => println!("{n}"),
        AirValue::Bool(b) => println!("{b}"),
        AirValue::Str(s) => println!("{s}"),
        other => println!("{other:?}"),
    }
}
